File: db4on/tutorial/db4o-tutorial-chapters/src/com/db4o/f1/chapter1/FirstStepsExample.cs
using System;
using System.IO;
using com.db4o;
using com.db4o.f1;

namespace com.db4o.f1.chapter1
{
	public class FirstStepsExample : Util
	{    
		public static void Main(string[] args)
		{
			File.Delete(Util.YapFileName);
			AccessDb4o();
			File.Delete(Util.YapFileName);
			ObjectContainer db = Db4o.OpenFile(Util.YapFileName);
			try
			{
				StoreFirstPilot(db);
				StoreSecondPilot(db);
				RetrieveAllPilots(db);
				RetrievePilotByName(db);
				RetrievePilotByExactPoints(db);
				UpdatePilot(db);
				DeleteFirstPilotByName(db);
				DeleteSecondPilotByName(db);
			}
			finally
			{
				db.Close();
			}
		}
        
		public static void AccessDb4o()
		{
			ObjectContainer db=Db4o.OpenFile(Util.YapFileName);
			try
			{
				// do something with db4o
			}
			finally
			{
				db.Close();
			}
		}
        
		public static void StoreFirstPilot(ObjectContainer db)
		{
			Pilot pilot1 = new Pilot("Michael Schumacher", 100);
			db.Set(pilot1);
			Console.WriteLine("Stored {0}", pilot1);
		}
    
		public static void StoreSecondPilot(ObjectContainer db)
		{
			Pilot pilot2 = new Pilot("Rubens Barrichello", 99);
			db.Set(pilot2);
			Console.WriteLine("Stored {0}", pilot2);
		}
    
		public static void RetrieveAllPilotQBE(ObjectContainer db) 
		{
			Pilot proto = new Pilot(null, 0);
			ObjectSet result = db.Get(proto);
			ListResult(result);
		}
    
		public static void RetrieveAllPilots(ObjectContainer db) 
		{
			ObjectSet result = db.Get(typeof(Pilot));
			ListResult(result);
		}
    
		public static void RetrievePilotByName(ObjectContainer db)
		{
			Pilot proto = new Pilot("Michael Schumacher", 0);
			ObjectSet result = db.Get(proto);
			ListResult(result);
		}
        
		public static void RetrievePilotByExactPoints(ObjectContainer db)
		{
			Pilot proto = new Pilot(null, 100);
			ObjectSet result = db.Get(proto);
			ListResult(result);
		}
    
		public static void UpdatePilot(ObjectContainer db)
		{
			ObjectSet result = db.Get(new Pilot("Michael Schumacher", 0));
			Pilot found = (Pilot)result.Next();
			found.AddPoints(11);
			db.Set(found);
			Console.WriteLine("Added 11 points for {0}", found);
			RetrieveAllPilots(db);
		}
    
		public static void DeleteFirstPilotByName(ObjectContainer db)
		{
			ObjectSet result = db.Get(new Pilot("Michael Schumacher", 0));
			Pilot found = (Pilot)result.Next();
			db.Delete(found);
			Console.WriteLine("Deleted {0}", found);
			RetrieveAllPilots(db);
		}
    
		public static void DeleteSecondPilotByName(ObjectContainer db)
		{
			ObjectSet result = db.Get(new Pilot("Rubens Barrichello", 0));
			Pilot found = (Pilot)result.Next();
			db.Delete(found);
			Console.WriteLine("Deleted {0}", found);
			RetrieveAllPilots(db);
		}
	}
}

File: db4on/tutorial/vb/db4o-tutorial-chapters/src/com/db4o/f1/chapter1/FirstStepsExample.vb
Imports System
Imports System.IO
Imports com.db4o

Namespace com.db4o.f1.chapter1
	Public Class FirstStepsExample
	Inherits Util
		Public Shared Sub Main(ByVal args As String())
			File.Delete(Util.YapFileName)
			AccessDb4o()
			File.Delete(Util.YapFileName)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(Util.YapFileName)
			Try
				StoreFirstPilot(db)
				StoreSecondPilot(db)
				RetrieveAllPilots(db)
				RetrievePilotByName(db)
				RetrievePilotByExactPoints(db)
				UpdatePilot(db)
				DeleteFirstPilotByName(db)
				DeleteSecondPilotByName(db)
			Finally
				db.Close()
			End Try
		End Sub

		Public Shared Sub AccessDb4o()
            Dim db As ObjectContainer = Db4oFactory.OpenFile(Util.YapFileName)
			Try
				' do something with db4o
			Finally
				db.Close()
			End Try
		End Sub

		Public Shared Sub StoreFirstPilot(ByVal db As ObjectContainer)
			Dim pilot1 As Pilot = New Pilot("Michael Schumacher", 100)
            db.Set(pilot1)
			Console.WriteLine("Stored {0}", pilot1)
		End Sub

		Public Shared Sub StoreSecondPilot(ByVal db As ObjectContainer)
			Dim pilot2 As Pilot = New Pilot("Rubens Barrichello", 99)
			db.[Set](pilot2)
			Console.WriteLine("Stored {0}", pilot2)
		End Sub

		Public Shared Sub RetrieveAllPilotQBE(ByVal db As ObjectContainer)
			Dim proto As Pilot = New Pilot(Nothing, 0)
			Dim result As ObjectSet = db.[Get](proto)
			ListResult(result)
		End Sub

		Public Shared Sub RetrieveAllPilots(ByVal db As ObjectContainer)
			Dim result As ObjectSet = db.[Get](GetType(Pilot))
			ListResult(result)
		End Sub

		Public Shared Sub RetrievePilotByName(ByVal db As ObjectContainer)
			Dim proto As Pilot = New Pilot("Michael Schumacher", 0)
			Dim result As ObjectSet = db.[Get](proto)
			ListResult(result)
		End Sub

		Public Shared Sub RetrievePilotByExactPoints(ByVal db As ObjectContainer)
			Dim proto As Pilot = New Pilot(Nothing, 100)
			Dim result As ObjectSet = db.[Get](proto)
			ListResult(result)
		End Sub

		Public Shared Sub UpdatePilot(ByVal db As ObjectContainer)
			Dim result As ObjectSet = db.[Get](New Pilot("Michael Schumacher", 0))
			Dim found As Pilot = DirectCast(result.[Next](), Pilot)
			found.AddPoints(11)
			db.[Set](found)
			Console.WriteLine("Added 11 points for {0}", found)
			RetrieveAllPilots(db)
		End Sub

		Public Shared Sub DeleteFirstPilotByName(ByVal db As ObjectContainer)
			Dim result As ObjectSet = db.[Get](New Pilot("Michael Schumacher", 0))
			Dim found As Pilot = DirectCast(result.[Next](), Pilot)
			db.Delete(found)
			Console.WriteLine("Deleted {0}", found)
			RetrieveAllPilots(db)
		End Sub

		Public Shared Sub DeleteSecondPilotByName(ByVal db As ObjectContainer)
			Dim result As ObjectSet = db.[Get](New Pilot("Rubens Barrichello", 0))
			Dim found As Pilot = DirectCast(result.[Next](), Pilot)
			db.Delete(found)
			Console.WriteLine("Deleted {0}", found)
			RetrieveAllPilots(db)
		End Sub

	End Class
End Namespace
